pytools/__init__.py

```diff
@@ -1438,23 +1453,42 @@ def generate_nonnegative_integer_tuples_summing_to_at_most(n, length):
 generate_positive_integer_tuples_below = generate_nonnegative_integer_tuples_below
 
 
-def _pos_and_neg_adaptor(tuple_iter):
+def _pos_and_neg_adaptor(
+        tuple_iter: Iterator[tuple[int, ...]]
+    ) -> Iterator[tuple[int, ...]]:
     for tup in tuple_iter:
         nonzero_indices = [i for i in range(len(tup)) if tup[i] != 0]
         for do_neg_tup in generate_nonnegative_integer_tuples_below(
                 2, len(nonzero_indices)):
+
             this_result = list(tup)
             for index, do_neg in enumerate(do_neg_tup):
                 if do_neg:
                     this_result[nonzero_indices[index]] *= -1
+
             yield tuple(this_result)
 
 
-def generate_all_integer_tuples_below(n, length, least_abs=0):
+def generate_all_integer_tuples_below(
+        n: int, length: int, least_abs: int = 0
+    ) -> Iterator[tuple[int, ...]]:
     return _pos_and_neg_adaptor(generate_nonnegative_integer_tuples_below(
         n, length, least_abs))
 
 
+@overload
+def generate_permutations(original: list[T]) -> Iterator[list[T]]: ...
+
+@overload
+def generate_permutations(original: tuple[T, ...]) -> Iterator[tuple[T, ...]]: ...
+
+@overload
+def generate_permutations(original: str) -> Iterator[str]: ...
+
+@overload
+def generate_permutations(original: bytes) -> Iterator[bytes]: ...
+
+
 def generate_permutations(original):
```

Comment:
Shouldn't this also be typed?

Reply:
I'm not sure.. my understanding of the docs https://docs.python.org/3/library/typing.html#typing.overload was that the base function doesn't get typed?